File: color.py
from util.database.color import add_color, remove_color, update_color, fetch_color
from util.database.database import Database


class Color():
    def __init__(self, id, color_sequence_id: int = None, position: int = None,
                 red: int = None, green: int = None, blue: int = None):
        self.id = id
        
        if color_sequence_id != None and red != None and green != None and blue != None:  # Position not included, because it is often added later
            self.color_sequence_id = color_sequence_id
            self.position = position
            self.red = red
            self.green = green
            self.blue = blue
        
        elif not self.id:
            color_data = fetch_color(id)
            
            self.color_sequence_id = color_data['color_sequence_id']
            self.position = color_data['position']
            self.red = color_data['red']
            self.green = color_data['green']
            self.blue = color_data['blue']

    def sync_changes_to_db(self, new: bool = False) -> bool:
        if new:
            max_id = Database.fetchone_from_db('SELECT MAX(id) FROM colors', {})[0]
            
            self.id = 0
            if max_id != None:
                self.id = max_id + 1
            
            return add_color(self.id, self.color_sequence_id, self.position, self.red, self.green, self.blue)
        
        return update_color(self.id, self.color_sequence_id, self.position, self.red, self.green, self.blue)
    
    # def set_position(self, position: int):
    #     self.position = position

    #     return self.sync_changes_to_db()
            
    # def set_red(self, red: int):
    #     self.red = red

    #     return self.sync_changes_to_db()
            
    # def set_green(self, green: int):
    #     self.green = green

    #     return self.sync_changes_to_db()
            
    # def set_blue(self, blue: int):
    #     self.blue = blue

    #     return self.sync_changes_to_db()

File: src/Color.cpp
#include "Color.hpp"
#include "util/database/ColorDB.hpp"
#include "util/database/Database.hpp"

Color::Color(std::optional<int> id, 
             std::optional<int> color_sequence_id, 
             std::optional<int> position, 
             std::optional<int> red, 
             std::optional<int> green, 
             std::optional<int> blue) {
    if (id) {
        auto color_data = ColorDB::fetch_color(id.value());  // From ColorDB
        
        // Might be a problem here
        this->id = color_data.id;
        this->color_sequence_id = color_data.color_sequence_id;
        this->position = color_data.position;
        this->red = color_data.red;
        this->green = color_data.green;
        this->blue = color_data.blue;
    } else if (color_sequence_id && red && green && blue) {
        this->color_sequence_id = color_sequence_id.value();
        this->red = red.value();
        this->green = green.value();
        this->blue = blue.value();
    }
}

bool Color::sync_changes_to_db(bool is_new) {
    if (is_new) {
        auto response = Database::fetchone_from_db("SELECT MAX(id) FROM color_sequence_colors;", {});
        if (!response.empty()) {
            int max_id = std::stoi(response[0]);

            if (max_id < 0) {
                max_id = 0;
            }
            this->id = max_id + 1;
        } else {
            this->id = 0; // Start from 0 if no colors exist
        }
        
        return ColorDB::add_color(this->id, this->color_sequence_id, this->position, this->red, this->green, this->blue);
    } else {
        return ColorDB::update_color(this->id, this->color_sequence_id, this->position, this->red, this->green, this->blue);
    }
}
